drivers/gles3/rasterizer_canvas_gles3.cpp

```diff
@@ -1412,11 +1414,27 @@ void RasterizerCanvasGLES3::_render_batch(Light *p_lights, uint32_t p_index, Ren
 				GLuint vertex_array_gl = 0;
 				GLuint index_array_gl = 0;
 
-				uint64_t input_mask = RS::ARRAY_FORMAT_VERTEX | RS::ARRAY_FORMAT_COLOR | RS::ARRAY_FORMAT_TEX_UV; // 2D meshes always use the same vertex format.
+				uint64_t vertex_input_mask = RS::ARRAY_FORMAT_VERTEX | RS::ARRAY_FORMAT_COLOR | RS::ARRAY_FORMAT_TEX_UV;
+				RID material = state.canvas_instance_batches[p_index].material;
+				GLES3::CanvasMaterialData *material_data = nullptr;
+				GLES3::CanvasShaderData *shader_data = nullptr;
+
+				if (material.is_valid()) {
+					material_data = static_cast<GLES3::CanvasMaterialData *>(material_storage->material_get_data(material, RS::SHADER_CANVAS_ITEM));
+				}
+				if (material_data) {
+					if (material_data->shader_data->version.is_valid() && material_data->shader_data->valid) {
+						shader_data = material_data->shader_data;
+						if (shader_data) {
+							vertex_input_mask = shader_data->vertex_input_mask;
+						}
+					}
+				}
```

Comment:
A couple of things here:
1. `material_data` is already available through the batch, It can be accessed with `state.canvas_instance_batches[p_index].material_data` so we can skip the first `if` statement
2. `vertex_input_mask` comes from the material. So instead of doing all these checks in the hot loop, we should do them when the batch is first created. Accordingly, I would add `vertex_input_mask` into the batch data structure and then cache it when caching `material` and `material_data`:

https://github.com/godotengine/godot/blob/6fea273ed3df7d4be9674d35aae698731fa823ea/drivers/gles3/rasterizer_canvas_gles3.cpp#L628-L629